src/basic/conf-virt.c

```diff
@@ -0,0 +1,290 @@
+                    CPUID_HYPERV_ISOLATION_TYPE_SNP) {
+                        return true;
+                }
+        }
+
+        return false;
+}
+
+static ConfidentialVirtualization
+detect_sev(void)
+{
+        uint32_t eax, ebx, ecx, edx;
+        uint64_t msrval;
+
+        eax = CPUID_GET_HIGHEST_FUNCTION;
+        ebx = ecx = edx = 0;
+
+        cpuid(&eax, &ebx, &ecx, &edx);
+
+        if (eax < CPUID_AMD_GET_ENCRYPTED_MEMORY_CAPABILITIES)
+                return CONFIDENTIAL_VIRTUALIZATION_NONE;
+
+        eax = CPUID_AMD_GET_ENCRYPTED_MEMORY_CAPABILITIES;
+        ebx = ecx = edx = 0;
+
+        cpuid(&eax, &ebx, &ecx, &edx);
+
+        /* bit 1 == CPU supports SEV feature
+         *
+         * Note, Azure blocks this CPUID leaf from its SEV-SNP
+         * guests, so we must fallback to probing the TPM which
+         * exposes a SEV-SNP attestation report as evidence.
+         */
+        if (!(eax & (1 << 1))) {
+                log_debug("No sev in CPUID, try hyperv CPUID");
+
+                if (detect_hyperv_sev()) {
+                        return CONFIDENTIAL_VIRTUALIZATION_SEV_SNP;
+                } else {
+                        log_debug("No hyperv CPUID");
+                }
+                return CONFIDENTIAL_VIRTUALIZATION_NONE;
+        }
+
+        msrval = msr(MSR_AMD64_SEV);
+
+        /* Test reverse order, since the SEV-SNP bit implies
+         * the SEV-ES bit, which implies the SEV bit */
+        if (msrval & (1 << 2)) {
+                return CONFIDENTIAL_VIRTUALIZATION_SEV_SNP;
+        } else if (msrval & (1 << 1)) {
+                return CONFIDENTIAL_VIRTUALIZATION_SEV_ES;
+        } else if (msrval & (1 << 0)) {
+                return CONFIDENTIAL_VIRTUALIZATION_SEV;
+        }
+        return CONFIDENTIAL_VIRTUALIZATION_NONE;
+}
+
+static ConfidentialVirtualization detect_tdx(void)
+{
+        uint32_t eax, ebx, ecx, edx;
+        char sig[13];
+
+        eax = CPUID_GET_HIGHEST_FUNCTION;
+        ebx = ecx = edx = 0;
+
+        cpuid(&eax, &ebx, &ecx, &edx);
+        log_debug("CPUID max function: %x %x %x %x", eax, ebx, ecx,edx);
+
+        if (eax < CPUID_INTEL_TDX_ENUMERATION)
+                return CONFIDENTIAL_VIRTUALIZATION_NONE;
+
+        memset(sig, 0, sizeof sig);
+        cpuid_leaf(CPUID_INTEL_TDX_ENUMERATION, sig, true);
+
+        if (memcmp(sig, CPUID_SIG_INTEL_TDX, sizeof(sig)) == 0)
+                return CONFIDENTIAL_VIRTUALIZATION_TDX;
+
+        return CONFIDENTIAL_VIRTUALIZATION_NONE;
+}
+
+static bool
+detect_hypervisor(void)
+{
+        uint32_t eax, ebx, ecx, edx;
+        bool is_hv;
+
+        eax = CPUID_PROCESSOR_INFO_AND_FEATURE_BITS;
+        ebx = ecx = edx = 0;
+
+        cpuid(&eax, &ebx, &ecx, &edx);
+
+        is_hv = ecx & CPUID_FEATURE_HYPERVISOR;
+
+        log_debug("CPUID is hypervisor: %s", is_hv ? "yes" : "no");
+        return is_hv;
+}
+
+
+ConfidentialVirtualization detect_confidential_virtualization(void) {
+        char sig[13];
+
```

Comment:
This is not going to change between invocations, so have a static cached var here to check it only once

Reply:
that sounds unnecessary to me. caching is for slow ops. I am pretty sure getting 12 bytes from the cpu itself is going to be a ton faster than reading them from memory, hence no reason to cache. it probably would be worse...

Reply:
it also opens an fd, it's not just a read from memory, no?

Reply:
ah right. you wanted to cache the ultimate result, which yes, makes more sense. given the location of the comment i misunderstood you and thought you wanted to cache sig[13]...

Reply:
yes I mean the end result of the check

Reply:
I'll follow the same caching pattern that 'detect_vm' uses.